plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Internal.hs

```diff
@@ -0,0 +1,124 @@
+                                                       SemanticTokens,
+                                                       type (|?) (InL))
+import           Prelude                              hiding (span)
+
+logWith :: (MonadIO m) => IdeState -> Priority -> String -> m ()
+logWith st prior = liftIO . logPriority (ideLogger st) prior . T.pack
+
+-----------------------
+---- the api
+-----------------------
+
+computeSemanticTokens :: IdeState -> NormalizedFilePath -> ExceptT PluginError Action SemanticTokens
+computeSemanticTokens state nfp = do
+    let dbg = logWith state Debug
+    dbg $ "Computing semantic tokens for: " <> show nfp
+    (RangeHsSemanticTokenTypes {tokens}, mapping) <- useWithStaleE GetSemanticTokens nfp
+    let rangeTokens = mapMaybe (\(span, name) -> (,name) <$> toCurrentRange mapping span) tokens
+    withExceptT PluginInternalError $ liftEither $ semanticTokenAbsoluteSemanticTokens rangeTokens
+
+semanticTokensFull :: PluginMethodHandler IdeState 'Method_TextDocumentSemanticTokensFull
+semanticTokensFull state _ param = do
+  nfp <- getNormalizedFilePathE (param ^. L.textDocument . L.uri)
+  items <- runActionE "SemanticTokens.semanticTokensFull" state $ computeSemanticTokens state nfp
+  return $ InL items
+
+-- | Defines the 'getSemanticTokensRule' function, compute semantic tokens for a Haskell source file.
+--
+-- This Rule collects information from various sources, including:
+--
+-- Imported name token type from Rule 'GetSemanticTokens'
+-- Local names token type from 'hieAst'
+-- Name locations from 'hieAst'
+-- Visible names from 'tmrRenamed'
+--
+-- It then combines this information to compute the semantic tokens for the file.
+--
+getSemanticTokensRule :: Recorder (WithPriority Log) -> Rules ()
+getSemanticTokensRule recorder =
+  define (cmapWithPrio LogShake recorder) $ \GetSemanticTokens nfp -> do
+    (TcModuleResult {..}) <- use_ TypeCheck nfp
+    (hscEnv -> hsc) <- use_ GhcSessionDeps nfp
+    (HAR {..}) <- use_ GetHieAst nfp
+    Just (_, ast) <- return $ listToMaybe $ Map.toList $ getAsts hieAst
+    let nameSet = nameGetter tmrRenamed
+    -- get imported name semantic map
+    importedNameSemanticMap <- liftIO $ foldrM (getTypeExclude (tcg_type_env tmrTypechecked) hsc) emptyNameEnv $ nameSetElemsStable nameSet
+    -- get current location from the old ones
+    let spanNames = hieAstSpanNames nameSet ast
+    let sMap = plusNameEnv_C (<>) importedNameSemanticMap $ mkLocalNameSemanticFromAst hieKind refMap
+    let rangeTokenType = extractSemanticTokensFromNames sMap spanNames
+    return ([], Just $ RangeHsSemanticTokenTypes rangeTokenType)
+    where
+        -- ignore one already in current module
+        getTypeExclude localMap env n nameMap
+            | Nothing <- lookupNameEnv localMap n =
+                do  tyThing <- lookupImported env n
+                    pure $ maybe nameMap (extendNameEnv nameMap n) (tyThing >>= tyThingSemantic)
```

Comment:
Nit: personally I prefer case expressions over `maybe` unless it's partially applied